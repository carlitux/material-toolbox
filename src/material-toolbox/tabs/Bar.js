// @flow
import * as React from 'react';
import cx from 'classnames';

import MDCTabBarFoundation from '@material/tabs/tab-bar/foundation';

import type Tab from './Tab';

// TODO: Only accept tabs as children
type Props = {
  className?: string,
  cssOnly: boolean,
  children?: React.Node,
  refreshAt: number,
  onChange?: (index: number, tab: Tab) => void,
  defaultActive?: number,
};

type State = {
  classNames: { [string]: boolean },
  tabIndicatorStyles: { [string]: any },
};

export default class TabBar extends React.Component<Props, State> {
  static defaultProps = {
    className: undefined,
    cssOnly: false,
    children: null,
    defaultActive: undefined,
    refreshAt: 200,
  };

  state = {
    classNames: {},
    tabIndicatorStyles: {},
  };

  componentDidMount() {
    if (!this.props.cssOnly) {
      this.createMDCTabBar_();
      if (
        this.props.defaultActive !== null &&
        this.props.defaultActive !== undefined &&
        this.mdcTabBar_
      ) {
        this.mdcTabBar_.switchToTabAtIndex(this.props.defaultActive, false);
      }

      setTimeout(() => {
        window.dispatchEvent(new Event('resize'));
      }, this.props.refreshAt);
    }
  }

  componentWillReceiveProps(props: Props) {
    if (this.props.cssOnly !== props.cssOnly) {
      if (props.cssOnly) {
        this.removeMDCTabBar_();
      } else {
        this.createMDCTabBar_();
      }
    }
  }

  componentWillUnmount() {
    this.removeMDCTabBar_();
  }

  mdcTabBar_: ?MDCTabBarFoundation;
  tabBar_: ?HTMLElement;
  tabIndicator: ?HTMLSpanElement;
  tabs_: Array<Tab> = [];

  createMDCTabBar_() {
    this.mdcTabBar_ = new MDCTabBarFoundation({
      addClass: className =>
        this.setState(state => ({
          classNames: { ...state.classNames, [className]: true },
        })),
      removeClass: className =>
        this.setState(state => ({
          classNames: { ...state.classNames, [className]: undefined },
        })),
      registerResizeHandler: handler =>
        window.addEventListener('resize', handler),
      deregisterResizeHandler: handler =>
        window.removeEventListener('resize', handler),
      getOffsetWidth: () => this.tabBar_ && this.tabBar_.offsetWidth,
      setStyleForIndicator: (propertyName, value) =>
        this.setState(state => ({
          tabIndicatorStyles: {
            ...state.tabIndicatorStyles,
            [propertyName]: value,
          },
        })),
      notifyChange: console.log,
      getOffsetWidthForIndicator: () =>
        this.tabIndicator && this.tabIndicator.offsetWidth,
      getNumberOfTabs: () => React.Children.count(this.props.children),
      isTabActiveAtIndex: index =>
        this.tabs_[index] && this.tabs_[index].props.active,
      isDefaultPreventedOnClickForTabAtIndex: index =>
        this.tabs_[index].preventDefaultOnClick,
      setPreventDefaultOnClickForTabAtIndex: (index, preventDefaultOnClick) => {
        this.tabs_[index].preventDefaultOnClick = preventDefaultOnClick;
      },
      measureTabAtIndex: index =>
        this.tabs_[index] && this.tabs_[index].measureSelf(),
      getComputedWidthForTabAtIndex: index =>
        this.tabs_[index] && this.tabs_[index].computedWidth,
      getComputedLeftForTabAtIndex: index =>
        this.tabs_[index] && this.tabs_[index].computedLeft,
    });

    this.mdcTabBar_.init();
  }

  removeMDCTabBar_() {
    if (this.mdcTabBar_) {
      this.mdcTabBar_.destroy();
      delete this.mdcTabBar_;
    }
  }

  handleNotify = ({ tab }: { tab: Tab }) => {
    const index = this.tabs_.indexOf(tab);

    if (this.mdcTabBar_) {
      this.mdcTabBar_.switchToTabAtIndex(index, false);
    }

    if (this.props.onChange) {
      this.props.onChange(index, tab);
    }
  };

  render() {
    const {
      defaultActive,
      refreshAt,
      className,
      cssOnly,
      children,
      ...rest
    } = this.props;
    return (
      <nav
        {...rest}
        role="tablist" // eslint-disable-line
        className={cx(className, 'mdc-tab-bar', this.state.classNames)}
        ref={element => {
          this.tabBar_ = element;
        }}>
        {React.Children.map(children, (child, index) => (
          <child.type
            ref={element => {
              this.tabs_[index] = element;
            }}
            {...child.props}
            cssOnly_={cssOnly}
            onNotify_={this.handleNotify}
          />
        ))}
        {!cssOnly && (
          <span
            ref={element => {
              this.tabIndicator = element;
            }}
            className="mdc-tab-bar__indicator"
            style={this.state.tabIndicatorStyles}
          />
        )}
      </nav>
    );
  }
}
