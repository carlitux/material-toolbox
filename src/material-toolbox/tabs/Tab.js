// @flow
import * as React from 'react';
import cx from 'classnames';
import MDCTabFoundation from '@material/tabs/tab/foundation';

type Props = {
  className?: string,
  children?: React.Node,
  onClick?: (evt: Event) => void,
  onKeyDown?: (evt: Event) => void,
  active?: boolean,
  component?: React.ElementType | string,

  // Private to users
  cssOnly_?: boolean,
  onNotify_?: ({ tab: any }) => void,
};

type State = {
  classNames: { [string]: boolean },
  handlers: { [string]: (evt: Event) => void },
};

// TODO: Add ripple
export default class Tab extends React.PureComponent<Props, State> {
  static defaultProps = {
    className: undefined,
    cssOnly_: false,
    active: false,
    children: null,
  };

  state = {
    classNames: {},
    handlers: {},
  };

  componentDidMount() {
    if (!this.props.cssOnly_) {
      this.createMDCTab_();
    }
  }

  componentWillReceiveProps(props: Props) {
    if (this.props.cssOnly_ !== props.cssOnly_) {
      if (props.cssOnly_) {
        this.removeMDCTab_();
      } else {
        this.createMDCTab_();
      }
    }

    if (this.props.active !== props.active && this.mdcTab_) {
      this.mdcTab_.isActive(props.active);
    }
  }

  componentWillUnmount() {
    this.removeMDCTab_();
  }

  get computedWidth(): ?number {
    return this.mdcTab_ && this.mdcTab_.getComputedWidth();
  }

  get computedLeft(): ?number {
    return this.mdcTab_ && this.mdcTab_.getComputedLeft();
  }

  get preventDefaultOnClick(): ?boolean {
    return this.mdcTab_ && this.mdcTab_.preventsDefaultOnClick();
  }

  set preventDefaultOnClick(preventDefaultOnClick: boolean) {
    if (this.mdcTab_) {
      this.mdcTab_.setPreventDefaultOnClick(preventDefaultOnClick);
    }
  }

  measureSelf() {
    if (this.mdcTab_) {
      this.mdcTab_.measureSelf();
    }
  }

  mdcTab_: ?MDCTabFoundation;
  tab_: ?HTMLAnchorElement;

  handleClick = (evt: Event) => {
    if (this.state.handlers.click) {
      this.state.handlers.click(evt);
    }

    if (this.props.onClick) {
      this.props.onClick(evt);
    }
  };

  handleKeyDown = (evt: Event) => {
    if (this.state.handlers.keydown) {
      this.state.handlers.keydown(evt);
    }

    if (this.props.onKeyDown) {
      this.props.onKeyDown(evt);
    }
  };

  createMDCTab_() {
    this.mdcTab_ = new MDCTabFoundation({
      addClass: className =>
        this.setState(state => ({
          classNames: { ...state.classNames, [className]: true },
        })),
      removeClass: className =>
        this.setState(state => ({
          classNames: { ...state.classNames, [className]: undefined },
        })),
      registerInteractionHandler: (type, handler) =>
        this.setState(state => ({
          handlers: { ...state.handlers, [type]: handler },
        })),
      deregisterInteractionHandler: type =>
        this.setState(state => ({
          handlers: { ...state.handlers, [type]: undefined },
        })),
      getOffsetWidth: () => this.tab_ && this.tab_.offsetWidth,
      getOffsetLeft: () => this.tab_ && this.tab_.offsetLeft,
      notifySelected: () => {
        if (this.props.onNotify_) {
          this.props.onNotify_({ tab: this });
        }
      },
    });

    this.mdcTab_.init();
  }

  removeMDCTab_() {
    if (this.mdcTab_) {
      this.mdcTab_.destroy();
      delete this.mdcTab_;
    }
  }

  render() {
    const {
      className,
      onNotify_,
      cssOnly_,
      active,
      children,
      component: Component,
      ...rest
    } = this.props;
    return (
      <a
        {...rest}
        role="tab"
        onClick={this.handleClick}
        onKeyDown={this.handleKeyDown}
        tabIndex="0"
        className={cx(
          className,
          'mdc-tab',
          { 'mdc-tab--active': active },
          this.state.classNames,
        )}
        ref={element => {
          this.tab_ = element;
        }}>
        {children}
        {cssOnly_ && <span className="mdc-tab__indicator" />}
      </a>
    );
  }
}
