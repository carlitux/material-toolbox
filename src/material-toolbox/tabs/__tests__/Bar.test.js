import React from 'react';
import { shallow } from 'enzyme';
import Bar from '../Bar';
import Tab from '../Tab';

describe('component::Bar', () => {
  it('Should match snapshoot with cssOnly', () => {
    const component = shallow(
      <Bar cssOnly>
        <Tab>Tab 1</Tab>
        <Tab active>Tab 2</Tab>
      </Bar>,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot', () => {
    const component = shallow(
      <Bar className="extra-class">
        <Tab>Tab 1</Tab>
        <Tab active>Tab 2</Tab>
      </Bar>,
    );
    expect(component).toMatchSnapshot();
  });
});
