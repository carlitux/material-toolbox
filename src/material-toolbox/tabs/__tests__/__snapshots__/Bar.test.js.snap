// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::Bar Should match snapshoot 1`] = `
<nav
  className="extra-class mdc-tab-bar mdc-tab-bar-upgraded"
  role="tablist"
>
  <Tab
    active={false}
    cssOnly_={false}
    key=".0"
    onNotify_={[Function]}
  >
    Tab 1
  </Tab>
  <Tab
    active={true}
    cssOnly_={false}
    key=".1"
    onNotify_={[Function]}
  >
    Tab 2
  </Tab>
  <span
    className="mdc-tab-bar__indicator"
    style={Object {}}
  />
</nav>
`;

exports[`component::Bar Should match snapshoot with cssOnly 1`] = `
<nav
  className="mdc-tab-bar"
  role="tablist"
>
  <Tab
    active={false}
    cssOnly_={true}
    key=".0"
    onNotify_={[Function]}
  >
    Tab 1
  </Tab>
  <Tab
    active={true}
    cssOnly_={true}
    key=".1"
    onNotify_={[Function]}
  >
    Tab 2
  </Tab>
</nav>
`;
